sound/soc/sof/topology.c

```diff
@@ -1858,13 +1873,77 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 		offset += pdata->size;
 	}
 
+process:
 	process->size = ipc_data_size;
 	swidget->private = process;
 
 	ret = sof_ipc_tx_message(sdev->ipc, process->comp.hdr.cmd, process,
 				 ipc_size, r, sizeof(*r));
-	if (ret >= 0)
+
+	if (ret >= 0 && ipc_data_size)
 		return ret;
+
+	/* send control data with large message supported method */
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+
+		/* we only care about bytes control */
+		switch (widget->dobj.widget.kcontrol_type) {
+		case SND_SOC_TPLG_TYPE_MIXER:
+			sm = (struct soc_mixer_control *)kc->private_value;
+			scontrol = sm->dobj.private;
+			break;
+		case SND_SOC_TPLG_TYPE_BYTES:
+			sbe = (struct soc_bytes_ext *)kc->private_value;
+			scontrol = sbe->dobj.private;
+			break;
+		case SND_SOC_TPLG_TYPE_ENUM:
+			se = (struct soc_enum *)kc->private_value;
+			scontrol = se->dobj.private;
+			break;
+		default:
+			dev_err(sdev->dev, "error: unknown kcontrol type %d\n",
+				widget->dobj.widget.kcontrol_type);
+			ret = -EINVAL;
+			goto err;
+		}
+
+		/* don't include if there's no private data */
+		if (!scontrol->control_data->data)
+			continue;
+
+		/* make sure data is valid - data can be updated at runtime */
```

Comment:
is the comment correct? I thought we are in the part where we send stuff from topology, so what can be updated?

Reply:
this is an old comment so this is mystery to me also... so I know that data can be updated runtime but it doesn't too much connect to "validity". I'll fix it in separate commit.